This was supposed to be gone in sbt 1.0:
```
[test] /tmp/sbt_4ebeb8e9/xsbt/ConsoleInterface.scala:50: error: value setContextClassLoader is not a member of scala.tools.nsc.interpreter.Repl
[test]           intp.setContextClassLoader()
[test]                ^
[test] one error found
```

--- a/src/reflect/scala/reflect/internal/util/ScalaClassLoader.scala
+++ b/src/reflect/scala/reflect/internal/util/ScalaClassLoader.scala
@@ -13,18 +13,17 @@
 package scala
 package reflect.internal.util
 
-import java.lang.invoke.{ MethodHandles, MethodType }
+import java.lang.invoke.{MethodHandles, MethodType}
 
 import scala.language.implicitConversions
-import java.lang.{ ClassLoader => JClassLoader }
+import java.lang.{ClassLoader => JClassLoader}
 import java.lang.reflect.Modifier
-import java.net.{ URLClassLoader => JURLClassLoader }
+import java.net.{URLClassLoader => JURLClassLoader}
 import java.net.URL
 
-import scala.reflect.runtime.ReflectionUtils.{ show, unwrapHandler }
-import ScalaClassLoader._
-import scala.util.control.Exception.{ catching }
-import scala.reflect.{ ClassTag, classTag }
+import scala.reflect.runtime.ReflectionUtils.{show, unwrapHandler}
+import scala.util.control.Exception.catching
+import scala.reflect.{ClassTag, classTag}
 
 trait HasClassPath {
   def classPathURLs: Seq[URL]
@@ -36,11 +35,11 @@ trait HasClassPath {
 trait ScalaClassLoader extends JClassLoader {
   /** Executing an action with this classloader as context classloader */
   def asContext[T](action: => T): T = {
-    val saved = contextLoader
+    import ScalaClassLoader.setContext
+    val saved = Thread.currentThread.getContextClassLoader
     try { setContext(this) ; action }
     finally setContext(saved)
   }
-  def setAsContext(): Unit = { setContext(this) }
 
   /** Load and link a class with this classloader */
   def tryToLoadClass[T <: AnyRef](path: String): Option[Class[T]] = tryClass(path, initialize = false)
@@ -126,13 +125,7 @@ object ScalaClassLoader {
   }
   def contextLoader = apply(Thread.currentThread.getContextClassLoader)
   def appLoader     = apply(JClassLoader.getSystemClassLoader)
-  def setContext(cl: JClassLoader) =
-    Thread.currentThread.setContextClassLoader(cl)
-  def savingContextLoader[T](body: => T): T = {
-    val saved = contextLoader
-    try body
-    finally setContext(saved)
-  }
+  def setContext(cl: JClassLoader) = Thread.currentThread.setContextClassLoader(cl)
 
   class URLClassLoader(urls: Seq[URL], parent: JClassLoader)
       extends JURLClassLoader(urls.toArray, parent)

--- a/src/repl-frontend/scala/tools/nsc/Interpreter.scala
+++ b/src/repl-frontend/scala/tools/nsc/Interpreter.scala
@@ -111,7 +111,6 @@ class InterpreterLoop {
             override protected def newCompiler(settings: Settings, reporter: Reporter) =
               super.newCompiler(compilerSettings, reporter)
           }
-          interpreter.setContextClassLoader()
         }
 
         // for 2.8 compatibility

--- a/src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala
+++ b/src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala
@@ -982,7 +982,7 @@ class ILoop(config: ShellConfig, inOverride: BufferedReader = null,
   /** Start an interpreter with the given settings.
    *  @return true if successful
    */
-  def run(interpreterSettings: Settings): Boolean = savingContextLoader {
+  def run(interpreterSettings: Settings): Boolean = {
     if (!batchMode) printWelcome()
 
     in = defaultIn

--- a/src/repl/scala/tools/nsc/interpreter/IMain.scala
+++ b/src/repl/scala/tools/nsc/interpreter/IMain.scala
@@ -356,9 +356,6 @@ class IMain(val settings: Settings, parentClassLoaderOverride: Option[ClassLoade
       _runtimeClassLoader
     })
 
-  // Set the current Java "context" class loader to this interpreter's class loader
-  override def setContextClassLoader() = classLoader.setAsContext()
-
   def allDefinedNames: List[Name]  = exitingTyper(replScope.toList.map(_.name).sorted)
   def unqualifiedIds: List[String] = allDefinedNames.map(_.decode).sorted
 

--- a/src/repl/scala/tools/nsc/interpreter/Interface.scala
+++ b/src/repl/scala/tools/nsc/interpreter/Interface.scala
@@ -58,10 +58,8 @@ trait ReplCore {
     */
   def bindValue(name: String, value: Any): Result = bind(name, value.asInstanceOf[AnyRef].getClass.getName, value)
 
-  /** Set the current Java "context" class loader to this interpreter's class loader
-    *
-    */
-  def setContextClassLoader(): Unit
+  @deprecated("The thread context classloader is now set and restored around execution of REPL line, this method is now a no-op.", since = "2.12.0")
+  final def setContextClassLoader() = () // Called from sbt-interface/0.12.4/src/ConsoleInterface.scala:39
 }
 
 /**

--- a/test/junit/scala/tools/nsc/interpreter/ScriptedTest.scala
+++ b/test/junit/scala/tools/nsc/interpreter/ScriptedTest.scala
@@ -99,4 +99,31 @@ class ScriptedTest {
     val err = "not found: value foo in def f = foo at line number 1 at column number 9"
     assertThrows[ScriptException](engine.compile("def f = foo"), _ == err)
   }
+
+   @Test def `restore classloader`(): Unit = {
+    val saved0 = Thread.currentThread.getContextClassLoader
+    try {
+      Thread.currentThread.setContextClassLoader(ClassLoader.getSystemClassLoader)
+      val saved = Thread.currentThread.getContextClassLoader
+      val engine = scripted
+      scripted.eval("42")
+      val now = Thread.currentThread.getContextClassLoader
+      assert(saved eq now)
+    } finally {
+      Thread.currentThread.setContextClassLoader(saved0)
+    }
+  }
+  @Test def `restore classloader script api`(): Unit = {
+    val saved0 = Thread.currentThread.getContextClassLoader
+    try {
+      Thread.currentThread.setContextClassLoader(ClassLoader.getSystemClassLoader)
+      val saved = Thread.currentThread.getContextClassLoader
+      val engine = new ScriptEngineManager().getEngineByName("scala")
+      assertNotNull(engine)
+      val now = Thread.currentThread.getContextClassLoader
+      assert(saved eq now)
+    } finally {
+      Thread.currentThread.setContextClassLoader(saved0)
+    }
+  }
 }

